build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -396,7 +396,8 @@ message Action {
   // immediately, rather than whenever the cache entry gets evicted.
   google.protobuf.Duration timeout = 6;
 
-  // If true, then the `Action`'s result cannot be cached.
+  // If true, then the `Action`'s result cannot be cached, and in-flight
+  // requests for the same `Action` may not be merged.
```

Comment:
This is hard to realise in the case of Buildbarn.

Buildbarn currently treats the action as a black box. It merely queues an execute request and forwards it to a worker. The worker is responsible for loading the action from the CAS. If this change is pushed through, Buildbarn's scheduler would need to be extended to access the CAS, which it currently doesn't.

Maybe we should move this into the ExecuteRequest to make it more readily available?

Reply:
I recall that we wanted to make it part of the Action in order to ensure that actions that differ on this field alone have different action keys. But now that I think about it, I don't really see the point -- I don't see the different keys preventing any bad scenarios in a way that a field in ExecuteRequest wouldn't. So I wouldn't be opposed to moving this (even though it would be a breaking change, would need to be done in steps, etc).
Am I forgetting some reason?

Reply:
Ed, just to make sure I understand: is the semantic clarification that in-flight actions may not be deduped problematic for you, or more generally that some actions are uncacheable at all (as defined within the Action)?

I agree with Ola that moving this field out of the Action would be reasonable, so if its existence is problematic for you that seems fine to me. (But yeah, breaking change). I'd prefer to not block this change on it however if this doesn't make things _harder_ for you.

Reply:
> Ed, just to make sure I understand: is the semantic clarification that in-flight actions may not be deduped problematic for you, or more generally that some actions are uncacheable at all (as defined within the Action)?

The latter. Right now Buildbarn's scheduler always deduplicates actions. It would be possible to disable this logic, but then it would be disabled globally; not on a per-action basis. Buildbarn's scheduler has no access to the CAS and thus not to the Action message.

Shall I send out a pull request to move this flag after your pull request lands?

Reply:
Yes please! Though I suggest syncing with @ola-rozenfeld on _how_ ... I think this will need to be a change destined for V3? It's a breaking change that I'm not sure offhand how to make in a forwards and backwards compatible way.

Reply:
I think it'll be manageable. We can add the field to `ExecuteRequest` as well, stating that:

- The one in `ExecuteRequest` is the canonical one going forward.
- For compatibility, clients should set both to true and false.
- For compatibility, servers should 'or' both of them together.

Reply:
With the idea being that Buildbarn will then be noncompliant with the spec (you'll only actually check the one in ExecuteRequest), but you think that'll be benign?

Reply:
I like this plan, but let's do this change in a separate PR, and have the discussion there as well, for records.

Reply:
Exactly. Buildbarn can easily check both of them to determine whether something needs to be inserted into the cache (which it already does for the current flag), but will only respect the flag in the ExecuteRequest to prevent merging of in-flight requests.

Reply:
SGTM then! And glad to hear we don't have to consider V3 just yet :).